storybookjs/storybook
# Core: Fix indexing errors by excluding node_modules stories

Closes #21414 https://github.com/storybookjs/storybook/issues/19446

## What I did

Exclude stories from `node_modules` from the story index. Starting in 7.0, we are including story files in the Storybook renderer packages, which are installed into `node_modules`. This causes errors for users whose `stories` globs are configured to match `node_modules`. Also matching `node_modules` is a huge performance hog.

Now if you want to include stories from `node_modules` you need to explicitly add that to your globs. This is described in my change to `MIGRATION.md`.

This is technically a breaking change but I expect it to break very few users, and the problems that it fixes were introduced in 7.0, which is still stabilizing. Hence I think it is safe to patch back.

## How to test

In a sandbox:

1. Add a file `node_modules/Dummy.jsx` containing the following content:

```jsx
import React from 'react'
const Dummy = () => <div>Dummy</div>
export default { title: 'Dummy', component: Dummy }
export const Default = {};
```

2. Update the `main.js` to include the pattern `../**/*.stories.@(js|jsx|ts|tsx)`.
3. Verify that `Dummy` does not appear in the sidebar
4. Replace the exclude pattern in `common-glob-options.ts` (added in this PR) from '**/node_modules/**'` to `'**/@storybook/**'`
5. Verify that `Dummy` appears at the top of the sidebar

Perform all of the above:
- [ ] In a vite sandbox
  - [ ] Normally
  - [ ] In SSv6 mode by setting `features.storyStoryV7 = true` and `features.storyStoreV7MdxErrors = false`
- [ ] In a webpack sandbox
  - [ ] Normally
  - [ ] In SSv6 mode
